osu.Game/OsuGame.cs

```diff
@@ -341,16 +339,28 @@ public void PresentBeatmap(BeatmapSetInfo beatmap, Predicate<BeatmapInfo> diffic
                     menuScreen.LoadToSolo();
 
                 // we might even already be at the song
-                if (Beatmap.Value.BeatmapSetInfo.Hash == databasedSet.Hash && difficultyCriteria(Beatmap.Value.BeatmapInfo))
+                if (Beatmap.Value.BeatmapSetInfo.Hash == databasedSet.Hash && (difficultyCriteria?.Invoke(Beatmap.Value.BeatmapInfo) ?? true))
                 {
                     return;
                 }
 
-                // Find first beatmap that matches our predicate.
-                var first = databasedSet.Beatmaps.Find(difficultyCriteria) ?? databasedSet.Beatmaps.First();
+                // Find beatmaps that match our predicate.
+                var beatmaps = databasedSet.Beatmaps.Where(b => difficultyCriteria?.Invoke(b) ?? true);
+
+                // Use all beatmaps if predicate matched nothing
+                if (!beatmaps.Any())
+                    beatmaps = databasedSet.Beatmaps;
+
+                // Try to select recommended beatmap
+                // This should give us a beatmap from current ruleset if there are any in our matched beatmaps
+                var selection = DifficultyRecommender.GetRecommendedBeatmap(beatmaps) ?? (
+                    // Fallback if a difficulty can't be recommended, maybe we are offline
+                    // First try to find a beatmap in current ruleset, otherwise use first beatmap
+                    beatmaps.FirstOrDefault(b => b.Ruleset.Equals(Ruleset.Value)) ?? beatmaps.First()
+                );
```

Comment:
This reads super awkwardly. I'd split into three lines using `??=`.

Reply:
This used to be split into multiple lines but that triggered CI errors. I'll see if using ``??=`` fixes that.